src/terminal/input/terminalInput.cpp

```diff
@@ -706,60 +662,58 @@ bool TerminalInput::HandleKey(const IInputEvent* const pInEvent)
     // These mappings will kick in no matter which modifiers are pressed and as such
     // must be checked last, or otherwise we'd override more complex key combinations.
     const auto mapping = _getKeyMapping(keyEvent, _inputMode.test(Mode::Ansi), _inputMode.test(Mode::CursorKey), _inputMode.test(Mode::Keypad));
-    if (_translateDefaultMapping(keyEvent, mapping, senderFunc))
+    if (const auto match = _searchKeyMapping(keyEvent, mapping))
     {
-        return true;
+        return MakeOutput(match->sequence);
     }
 
     // If all else fails we can finally try to send the character itself if there is any.
     if (keyEvent.GetCharData() != 0)
     {
-        _SendChar(keyEvent.GetCharData());
-        return true;
+        return _SendChar(keyEvent.GetCharData());
     }
 
-    return false;
+    return MakeUnhandled();
 }
 
-bool TerminalInput::HandleFocus(const bool focused) noexcept
+TerminalInput::OutputType TerminalInput::HandleFocus(const bool focused) const
 {
-    const auto enabled{ _inputMode.test(Mode::FocusEvent) };
-    if (enabled)
+    if (!_inputMode.test(Mode::FocusEvent))
     {
-        _SendInputSequence(focused ? L"\x1b[I" : L"\x1b[O");
+        return MakeUnhandled();
     }
-    return enabled;
+
+    return MakeOutput(focused ? L"\x1b[I" : L"\x1b[O");
 }
 
 // Routine Description:
 // - Sends the given character to the shell.
 // - Surrogate pairs are being aggregated by this function before being sent.
 // Arguments:
 // - ch: The UTF-16 character to send.
-void TerminalInput::_SendChar(const wchar_t ch)
+TerminalInput::OutputType TerminalInput::_SendChar(const wchar_t ch)
 {
+    StringType str;
+
     if (til::is_leading_surrogate(ch))
     {
         if (_leadingSurrogate.has_value())
         {
-            // we already were storing a leading surrogate but we got another one. Go ahead and send the
-            // saved surrogate piece and save the new one
-            const auto formatted = wil::str_printf<std::wstring>(L"%I32u", _leadingSurrogate.value());

```

Comment:
what in blazes? so we would literally just send the raw value of the leading surrogate as an integer?

Reply:
this is a behavior change, but i can't see anyone relying on it.
the code on the right just _sends_ the leading surrogate as a character instead.

Reply:
noting of course that this is the _broken_ case where we got a Lead + Lead

Reply:
Yeah, I was quite confused as well. I looked up where it came from and it's been there ever since it got introduced with the WPF PR.

Reply:
I made it so that it now discard invalid surrogate pairs entirely. It's rather unexpected that our input handling code would have to deal with unpaired surrogate pairs anyways. Discarding them is probably about as good as sending U+FFFD.